text/0032-upgrade-bootstrapper.md

```diff
@@ -0,0 +1,153 @@
+  config.toml file all to live in the same folder
+
+### Data migration
+
+We require to add a new `migrate-data` subcommand to `casper-node`.  We could provide a separate binary to be used
+exclusively for data and config migration, but it seems simpler to just make this a subcommand.  Simpler for the
+installer, for the `upgrade-bootstrapper` and simpler to code.
+
+Migration of any of the databases needs to take the following approach by iterating each value in the DB and doing:
+
+1. parse as the old type
+2. convert it to the new type
+3. store the new type, possibly under a different key
+4. if stored under a different key (i.e. it's not just an overwrite), then delete the old value
+
+We need to be able to resume migration in case it's killed while running.  This means we need to ensure that we either
+keep track of our progress by writing to disk after each individual item is successfully migrated, or else we ensure
+that parsing a new value as an old value fails, or else we ensure that running the conversion step on an
+already-converted item is a no-op.
+
+If data migration is going to take a prohibitively long time to execute, we should look to have the `upgrade-installer`
+perform _data_ migration only (not config migration) as soon as the installer provides the new `casper-node` binary,
+while the current `casper-node` is still running in validator mode.  However, this might result in a doubling or more of
+the disk space consumed by the databases, since the migrated data cannot be allowed to overwrite the existing data while
+the node is still running as a validator, and _this_ disk consumption might prove to be prohibitive.  Data migration
+would also still need to be executed again once the old `casper-node` stops due to reaching the upgrade activation
+point, since more data will likely have been written to the databases between then and when the data migration was
+previously completed.
+
+### Config migration
+
+Migration will also need to handle converting an old config file to a new one.  For this to work, we need to know which
+config values have been modified by the user from their default values.  This requires either default values to be
+provided, hard-coded in the node, or else the `migrate-data` subcommand will need to be passed not just the old and new
+config.toml files, but also a copy of each with their default values set.
+
+We moved away from hard-coding defaults in the code a while ago, but a challenge to the other approach is ensuring the
+user _copies_ the installed vanilla config rather than renaming or deleting it.  Two possible solutions are:
+1. have the installer create a read-only copy of the default config exclusively for the `upgrade-bootstrapper`'s use,
+   and a separate writable one for the user to modify/copy/move as they wish
+1. require that every config value is annotated with a standard comment specifying the default value
```

Comment:
```suggestion
2. require that every config value is annotated with a standard comment specifying the default value
```

Reply:
This isn't really required as the list is correctly displayed when rendered.  I tend to favour always using `1.` for each point as I don't have to renumber if I insert a point, although I favour explicit numbering if I'm explicitly referring to one point from another.  (I know I broke my own guidelines in the Data migration" section!)